Metroid Planets (Zebeth) support

UltiNaruto opened the pull request:
Done :
- Randomizes 42 items (including boss rewards on Kraid and Ridley)
- Random starting area (partially, requires some work to get the coordinates of the other starting locations to add)
- Goal (either x keys up to 9 or vanilla behavior so keys on Kraid and Ridley)

To do :
- Implement door lock modifications (not a priority)
- Implement elevator rando
- Maybe some cosmetics option (not a priority)
- Find someone interested in maintaining the logic DB (high priority)
- Multiworld support (way later)
- Figure out how to handle multiple vanilla locations for one item (not a priority)
- Add tests (will need some help for that)

Miepee commented:
some thoughts before i fully review
- you may wanna fix the current failing tests
- if you havent done so already, please run alll the images through [pnquant](https://pngquant.org/). dont feel afraid to toy around with its compression settings. Please force push after that, so that the original commit/images are not in the git history
- talk to planets/nestroid discord servers if you havent yet, if someone wants to co-maintain a logical DB

Miepee commented:
you also forgot to rebase i think. your logic is still in `json_data`, we moved that to `logic_database` a while ago.

UltiNaruto commented:
> some thoughts before i fully review
> 
> * you may wanna fix the current failing tests
> * if you havent done so already, please run alll the images through [pnquant](https://pngquant.org/). dont feel afraid to toy around with its compression settings. Please force push after that, so that the original commit/images are not in the git history
> * talk to planets/nestroid discord servers if you havent yet, if someone wants to co-maintain a logical DB

* I'll need some help for that
* I'll automate the process with a batch
* Definitely need to get to it

> you also forgot to rebase i think. your logic is still in `json_data`, we moved that to `logic_database` a while ago.

I rebased but didn't know that it changed

Miepee commented:
you can click on details, and then see backtraces / which tests failed:
![grafik](https://github.com/randovania/randovania/assets/38186597/5f2099cc-f140-4875-8bb3-765f33e62c7e)


All currently failing tests seem to be due to the json_data / logic_db mismatch. usually something:tm: creates a binary version of the db (.bin), which all the tests cant find

ThanatosGit commented:
It would also be nice to fix the mypy errors before reviewing ;)

UltiNaruto commented:
> you can click on details, and then see backtraces / which tests failed: ![grafik](https://user-images.githubusercontent.com/38186597/278862454-5f2099cc-f140-4875-8bb3-765f33e62c7e.png)
> 
> All currently failing tests seem to be due to the json_data / logic_db mismatch. usually something™️ creates a binary version of the db (.bin), which all the tests cant find

this one should be good to go

Miepee commented on randovania/games/planets_zebeth/gui/dialog/cosmetic_patches_dialog.py:
```diff
@@ -0,0 +1,41 @@
+        self.on_new_cosmetic_patches(current)
+        self.connect_signals()
+
+    def connect_signals(self):
```
mypy error

Miepee commented on randovania/games/planets_zebeth/gui/dialog/cosmetic_patches_dialog.py:
```diff
@@ -0,0 +1,41 @@
+        super().connect_signals()
+        # More signals here!
+
+    def on_new_cosmetic_patches(self, patches: PlanetsZebethCosmeticPatches):
```
mypy error

Miepee commented on randovania/games/planets_zebeth/gui/dialog/cosmetic_patches_dialog.py:
```diff
@@ -0,0 +1,41 @@
+    def cosmetic_patches(self) -> PlanetsZebethCosmeticPatches:
+        return self._cosmetic_patches
+
+    def reset(self):
```
mypy error

Miepee commented on randovania/games/planets_zebeth/gui/dialog/game_export_dialog.py:
```diff
@@ -0,0 +1,54 @@
+
+    @classmethod
+    def game_enum(cls):
+        return RandovaniaGame.METROID
```
wrong enum

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/planets_zebeth_goal_tab.py:
```diff
@@ -0,0 +1,77 @@
+    def uses_patches_tab(cls) -> bool:
+        return False
+
+    def _update_slider(self):
```
mypy error

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/planets_zebeth_goal_tab.py:
```diff
@@ -0,0 +1,77 @@
+        else:
+            self.keys_slider.setEnabled(True)
+
+    def _edit_config(self, call: Callable[[PlanetsZebethArtifactConfig], PlanetsZebethArtifactConfig]):
```
mypy error

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/planets_zebeth_goal_tab.py:
```diff
@@ -0,0 +1,77 @@
+        with self._editor as editor:
+            editor.set_configuration_field("artifacts", call(config.artifacts))
+
+    def _on_vanilla_boss_keys(self, value: bool):
```
mypy error

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/planets_zebeth_patches_tab.py:
```diff
@@ -0,0 +1,45 @@
+    from randovania.layout.preset import Preset
+
+
+class PresetPlanetsZebethPatches(PresetTab):
```
not sure if you have any planned patches yet, but may or may not wanna add a TODO for here

Miepee commented on randovania/games/planets_zebeth/layout/planets_zebeth_configuration.py:
```diff
@@ -0,0 +1,33 @@
+
+
+@dataclasses.dataclass(frozen=True)
+class PlanetsZebethConfiguration(BaseConfiguration):
```
may or may not wanna add a todo for planned patches.

Miepee commented on randovania/games/planets_zebeth/layout/planets_zebeth_cosmetic_patches.py:
```diff
@@ -0,0 +1,38 @@
+from randovania.lib import enum_lib
+
+
+class PlanetsZebethRoomGuiType(Enum):
```
with it being used 3 times now, i wonder whether we should have an issue that makes this a general class....

Miepee commented on randovania/games/planets_zebeth/layout/planets_zebeth_cosmetic_patches.py:
```diff
@@ -0,0 +1,38 @@
+
+@dataclasses.dataclass(frozen=True)
+class PlanetsZebethCosmeticPatches(BaseCosmeticPatches):
+    show_unexplored_map: bool = True
```
i'm curious what your reasoning is for making this an option.

UltiNaruto replied:
show_unexplored_map aka enable map from start for other games

Miepee replied:
I can read that, yes. My question was, why is this an option, instead of forced?

UltiNaruto replied:
why would it be forced when other games like Prime 1 and 2 give us an option?

Miepee replied:
I completely forgot about this, whoops. Nevermind.

Miepee commented on randovania/games/planets_zebeth/layout/planets_zebeth_cosmetic_patches.py:
```diff
@@ -0,0 +1,38 @@
+    show_room_names: PlanetsZebethRoomGuiType = PlanetsZebethRoomGuiType.WITH_FADE
+
+    @classmethod
+    def game(cls):
```
mypy error

Miepee commented on randovania/games/planets_zebeth/generator/pool_creator.py:
```diff
@@ -0,0 +1,63 @@
+    from randovania.layout.base.base_configuration import BaseConfiguration
+
+BOSS_KEY_CATEGORY = pickup_category.PickupCategory(
+    name="key", long_name="Boss Key", hint_details=("some ", "keys"), hinted_as_major=False, is_key=True
```
should be "some key". singular, not plural.

Miepee commented on randovania/games/planets_zebeth/presets/starter_preset.rdvpreset:
```diff
@@ -0,0 +1,112 @@
+{
+    "schema_version": 58,
```
we're at a way higher version now.

Miepee commented on randovania/games/prime1/pickup_database/pickup-database.json:
```diff
@@ -214,7 +217,9 @@
             "model_name": "Combat Visor",
-            "offworld_models": {},
+            "offworld_models": {
+                "planets_zebeth": "spr_ITEM_Sensor_Visor"
```
Should ask toaster what he thinks about this.

Miepee replied:
toaster didn't care, resolving

Miepee commented on randovania/games/planets_zebeth/pickup_database/pickup-database.json:
```diff
@@ -0,0 +1,298 @@
+        "Missile Launcher": {
+            "pickup_category": "missile",
+            "broad_category": "missile_related",
+            "model_name": "spr_ITEM_Missile",
```
missile launcher and missile ammo having the same model sounds like it can very much lead to confusion

UltiNaruto replied:
Will be handled when multiworld update is worked on

Miepee commented on randovania/games/planets_zebeth/exporter/game_exporter.py:
```diff
@@ -0,0 +1,44 @@
+        """
+        return False
+
+    def _do_export_game(
```
mypy error

Miepee commented:
Looking at the image in-programm, I do feel like the planets/zebeth text should be slightly bigger.
![grafik](https://github.com/randovania/randovania/assets/38186597/016eeb88-6311-481d-a1a1-b24ff0024870)

Miepee commented on randovania/games/planets_zebeth/game_data.py:
```diff
@@ -0,0 +1,79 @@
+        (
+            "Which versions of Metroid Planets are supported?",
+            "Only version 1.27g is supported. "
+            "Later versions are embedding code in the executable"
```
```suggestion
            "Later versions are embedding code in the executable "
```

Miepee commented on randovania/gui/ui_files/games_tab_planets_zebeth_widget.ui:
```diff
@@ -0,0 +1,179 @@
+      <item>
+       <widget class="QLabel" name="intro_label">
+        <property name="text">
+         <string>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
```
Something is very weird with this text
![grafik](https://github.com/randovania/randovania/assets/38186597/7f56b5b8-fb92-4098-bbae-49134181eefe)
For comparison
![grafik](https://github.com/randovania/randovania/assets/38186597/ef7c82c1-4c58-401f-9299-83810b2fcd28)

UltiNaruto replied:
Oops forgot to update the boss keys thing in this tab

Miepee replied:
i was more referring to it being way bolder for some reason.

UltiNaruto replied:
I haven't changed the font

Miepee replied:
well something is clearly different tho :/

Miepee replied:
looked into it, for some reason you have a bunch of extra styles defined in there.
If you're using QT Designer, you can right click the text -> change formatted text. (i think changing the file via a text editor might work too)
you should remove every tag/style that ***isn't*** one of the following
- html
- head
- body
- p
- a href="tab://NAME" (for faq/difference links)
- span style=" text-decoration: underline; color:#007af4;" (for faq/difference links)

Miepee commented on randovania/gui/ui_files/games_tab_planets_zebeth_widget.ui:
```diff
@@ -0,0 +1,179 @@
+        <item>
+         <widget class="QLabel" name="differences_label">
+          <property name="text">
+           <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Randovania makes some changes to the original game in order to improve the game experience or to simply fix bugs in the original game.&lt;br/&gt;Many changes are optional and can be disabled in the options Randovania provides, but the following are &lt;span style=&quot; font-weight:600;&quot;&gt;always&lt;/span&gt; there:&lt;/p&gt;& ...
```
this text reads a bit funny.  "the following are always there" and then only lists one :p
 From head, you should mention the slight missile related changes with killing kriad/redliy

Miepee commented on randovania/games/planets_zebeth/presets/starter_preset.rdvpreset:
```diff
@@ -0,0 +1,112 @@
+                "Hi-Jump Boots": {
+                    "num_shuffled_pickups": 1
+                },
+                "Ice Beam": {
```
![grafik](https://github.com/randovania/randovania/assets/38186597/d545f964-f0c9-40d3-b860-782111821214)
Ice beam seems weird. this should look like this instead:
![grafik](https://github.com/randovania/randovania/assets/38186597/fd11628c-0381-416e-9025-bed62b2f78e3)

UltiNaruto replied:
what if you want to start with ice beam?

Miepee replied:
thats unrelated. Starter preset doesnt start with ice beam, and thus should default to "2 shuffled copies".
if you want to *modify* starter preset to start with ice beam,then you can of course always change it to "custom".

UltiNaruto replied:
if you do like on the second screenshot it shuffles only one ice beam

Miepee replied:
that sounds like you have a mistake in your implementation? echoes, dread and msr do this for some items, and it does shuffle all of them.

Miepee replied:
Was a gui bug, will be fixed later.

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/__init__.py:
```diff
@@ -0,0 +1,28 @@
+        PresetLocationPool,
+        PresetPlanetsZebethGoal,
+        MetroidPresetItemPool,
+        PresetPlanetsZebethPatches,
+        PresetStartingArea,
```
I would reverse these two

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/__init__.py:
```diff
@@ -0,0 +1,28 @@
+        PresetTrickLevel,
+        PresetLocationPool,
+        PresetPlanetsZebethGoal,
+        MetroidPresetItemPool,
```
I would add PresetPatcherEnergy

UltiNaruto commented on randovania/games/dread/pickup_database/pickup-database.json:
```diff
@@ -345,7 +351,9 @@
             "model_name": "powerup_sonar",
-            "offworld_models": {},
+            "offworld_models": {
+                "planets_zebeth": "spr_ITEM_Sensor_Visor"
```
It's a visor that shows hidden items and yeah it's a novus thing

UltiNaruto commented on randovania/games/dread/pickup_database/pickup-database.json:
```diff
@@ -674,6 +695,7 @@
             "model_name": "item_missiletankplus",
             "offworld_models": {
                 "am2r": "sItemSuperMissile",
+                "planets_zebeth": "spr_ITEM_Big_Missile",
```
It's a custom sprite with a green box instead of the red box around a normal missile

Miepee replied:
right, so ammo, thanks.

UltiNaruto commented on randovania/games/planets_zebeth/assets/hash_words.json:
```diff
@@ -0,0 +1,68 @@
+[
```
brick is the name of some tiles (same for bubble)
bridge is the bridge at Entrance to Tourian
the last 3 ones unsure

Miepee replied:
internal name or actually public name?

UltiNaruto commented on randovania/games/planets_zebeth/generator/pool_creator.py:
```diff
@@ -0,0 +1,63 @@
+    from randovania.game_description.resources.resource_database import ResourceDatabase
+    from randovania.layout.base.base_configuration import BaseConfiguration
+
+BOSS_KEY_CATEGORY = pickup_category.PickupCategory(
```
in that case maybe call it KEY_CATEGORY

Miepee replied:
i was more referring to the name right after actually. `long_name="Boss Key"`

Miepee commented on randovania/gui/ui_files/games_tab_planets_zebeth_widget.ui:
```diff
@@ -30,15 +30,12 @@
-&lt;p style=&quot; margin-top:12px; margin-bottom:12px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;For more details, check out the &lt;a href=&quot;tab://FAQ&quot;&gt;&lt;span style=&quot; text-decoration: underline; color:#007af4;&quot;&gt;FAQ&lt;/span&gt;&lt;/a&gt; and &lt;a href=&quot;tab://Differences&quot;&gt;&lt;span style=&quot; text-decoration: underl ...
-&lt;p style=&quot; margin-top:12px; margin-bottom:12px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;To get started, use the Quick Generate button to generate a game using the default settings!&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
+         <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;
+&lt;p&gt;Traverse Zebes and its depths and kill Kraid and Ridley to fight Mother Brain and escape Zebes.&lt;/p&gt;
```
This sentence reads a tad weird with how many `and`s it contains. 
`Traverse the depths of Zebes, to kill Kraid and Ridley in order to fight Mother Brain and Escape Zebes.`?

UltiNaruto replied:
it's outdated btw

Miepee replied:
comment still applies, even the current text contains too many `and`s for my liking.

UltiNaruto replied:
Changed it a bit

UltiNaruto replied:
![image](https://github.com/randovania/randovania/assets/14176397/76089eb6-ccc7-4406-90b9-206f0a9a7598)

Miepee replied:
would still use `Traverse the depths of Zebes` TBH

UltiNaruto replied:
Done

Miepee commented on randovania/gui/ui_files/games_tab_planets_zebeth_widget.ui:
```diff
@@ -30,15 +30,12 @@
+         <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;
+&lt;p&gt;Traverse Zebes and its depths and kill Kraid and Ridley to fight Mother Brain and escape Zebes.&lt;/p&gt;
+&lt;p&gt;Two statues will block the path to Tourian, unless you killed both Kraid and Ridley. You can find them in their own lair.&lt;/p&gt;
+&lt;p&gt;Some hints are added to locate different pickups but only ice beam and missiles are required.&lt;/p&gt;
```
Hints? since when? where?

Also uppercase item names please

UltiNaruto replied:
Same comment as above. I haven't chosen yet for hints. Though I got an idea about it.

Miepee replied:
Then please don't mention hints here, until a proper design for them was discussed.

Miepee commented on randovania/gui/ui_files/preset_planets_zebeth_goal.ui:
```diff
@@ -102,12 +102,12 @@
          </property>
          <property name="text">
-          <string>Vanilla boss keys</string>
+          <string>Vanilla tourian keys</string>
```
Uppercase item please

UltiNaruto commented on randovania/games/planets_zebeth/generator/bootstrap.py:
```diff
@@ -0,0 +1,7 @@
+from randovania.resolver.bootstrap import Bootstrap
+
+
+class PlanetsZebethBootstrap(Bootstrap):
```
Not sure what I would add there so I decided not to add anything yet

Miepee replied:
hence the suggestion to add a todo, so no one forgets.

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/__init__.py:
```diff
@@ -0,0 +1,30 @@
+    from randovania.interface_common.preset_editor import PresetEditor
+
+
+def preset_tabs(editor: PresetEditor, window_manager: WindowManager):
```
mypy error

Miepee replied:
mypy error

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/planets_zebeth_goal_tab.py:
```diff
@@ -0,0 +1,77 @@
+            editor.set_configuration_field("artifacts", call(config.artifacts))
+
+    def _on_vanilla_tourian_keys(self, value: bool) -> None:
+        def edit(config: PlanetsZebethArtifactConfig):
```
mypy error

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/planets_zebeth_goal_tab.py:
```diff
@@ -0,0 +1,77 @@
+        self._edit_config(edit)
+        self._update_slider()
+
+    def _on_keys_slider_changed(self):
```
mypy error

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/planets_zebeth_goal_tab.py:
```diff
@@ -0,0 +1,77 @@
+        plural = "s" if self.keys_slider.value() > 1 else ""
+        self.keys_slider_label.setText(f"{self.keys_slider.value()} Key{plural}")
+
+        def edit(config: PlanetsZebethArtifactConfig):
```
mypy error

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/planets_zebeth_goal_tab.py:
```diff
@@ -0,0 +1,77 @@
+
+        self._edit_config(edit)
+
+    def on_preset_changed(self, preset: Preset):
```
mypy error

Miepee reviewed (commented):
please fix the resolver failing. the etank in upper right brinstar is not reachable trickless.

Miepee commented on randovania/games/planets_zebeth/game_data.py:
```diff
@@ -0,0 +1,79 @@
+from randovania.layout.preset_describer import GamePresetDescriber
+
+
+def _options():
```
mypy error

Miepee commented on randovania/games/planets_zebeth/game_data.py:
```diff
@@ -0,0 +1,79 @@
+    )
+
+
+def _patch_data_factory():
```
mypy error

Miepee commented on randovania/games/planets_zebeth/game_data.py:
```diff
@@ -0,0 +1,79 @@
+    return PlanetsZebethPatchDataFactory
+
+
+def _exporter():
```
mypy error

UltiNaruto commented on randovania/games/am2r/pickup_database/pickup-database.json:
```diff
@@ -419,6 +434,7 @@
             "broad_category": "missile_related",
             "model_name": "sItemSMissileLauncher",
             "offworld_models": {
+                "planets_zebeth": "spr_ITEM_Super_Missile",
```
Will be handled when multiworld update is worked on

UltiNaruto commented on randovania/games/am2r/pickup_database/pickup-database.json:
```diff
@@ -401,7 +415,8 @@
                 "prime2": "MissileLauncher",
-                "dread": "powerup_icemissile"
+                "dread": "powerup_icemissile",
+                "planets_zebeth": "spr_ITEM_Missile"
```
Will be handled when multiworld update is worked on

Miepee commented on randovania/games/dread/pickup_database/pickup-database.json:
```diff
@@ -268,6 +268,7 @@
             "model_name": "powerup_icemissile",
             "offworld_models": {
                 "am2r": "sItemIceMissilesDread",
+                "planets_zebeth": "spr_ITEM_Ice_Beam",
```
I would remove this and thus make it default to the generic offworld model for now.

Miepee commented on randovania/games/dread/pickup_database/pickup-database.json:
```diff
@@ -352,7 +353,8 @@
             "offworld_models": {
-                "am2r": "sItemPulseRadarDread"
+                "am2r": "sItemPulseRadarDread",
+                "planets_zebeth": "spr_ITEM_Sensor_Visor"
```
I would remove this and thus make it default to the generic offworld model for now.

Miepee commented on randovania/games/planets_zebeth/assets/hash_words.json:
```diff
@@ -0,0 +1,65 @@
+    "Nest",
+    "Norfair",
+    "Nova",
+    "Pirate",
```
If this "Pirate" is from "Space Pirate", then I don't see the "Space".

Miepee commented on randovania/games/planets_zebeth/assets/hash_words.json:
```diff
@@ -0,0 +1,65 @@
+[
+    "Acid",
+    "Atomic",
```
Metroid 1 doesn't feature atomics. I'd remove this.

Miepee commented on randovania/games/planets_zebeth/assets/hash_words.json:
```diff
@@ -0,0 +1,65 @@
+    "Gamet",
+    "Geruta",
+    "Hi-Jump",
+    "Hive",
```
Metroid 1 doesn't feature Hive's (assuming that this is referencing the hives for mellows).

Miepee commented on randovania/games/planets_zebeth/assets/hash_words.json:
```diff
@@ -0,0 +1,65 @@
+    "Vine",
+    "Wave",
+    "Waver",
+    "Worker",
```
Nestroid doesn't feature any workers to my knowledge.

Miepee commented on randovania/games/planets_zebeth/assets/hash_words.json:
```diff
@@ -0,0 +1,65 @@
+    "Wave",
+    "Waver",
+    "Worker",
+    "Yakchi",
```
What is this supposed to be?

UltiNaruto replied:
Same as above

Miepee replied:
Github just shows the context of the previous lines
Line in question is Yakchi.

Miepee commented on randovania/games/planets_zebeth/generator/pool_creator.py:
```diff
@@ -0,0 +1,63 @@
+from __future__ import annotations
```
This whole file should get a test.

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/__init__.py:
```diff
@@ -0,0 +1,30 @@
+    from randovania.interface_common.preset_editor import PresetEditor
+
+
+def preset_tabs(editor: PresetEditor, window_manager: WindowManager) -> [type]:
```
mypy error.

Miepee commented on randovania/games/planets_zebeth/gui/preset_settings/planets_zebeth_goal_tab.py:
```diff
@@ -0,0 +1,77 @@
+from __future__ import annotations
```
This whole file should get tested

Miepee commented on randovania/games/planets_zebeth/layout/preset_describer.py:
```diff
@@ -0,0 +1,56 @@
+from __future__ import annotations
```
This file should get tested.

Miepee commented on randovania/games/prime1/pickup_database/pickup-database.json:
```diff
@@ -135,7 +135,8 @@
             "offworld_models": {
-                "am2r": "sItemWaveBeamPrime"
+                "am2r": "sItemWaveBeamPrime",
+                "planets_zebeth": "spr_ITEM_Wave_Beam"
```
IIRC dunc or toaster didn't like it if beams were only partially represented. I.e. if no plasma beam exists, you should probably remove the entries for wave and ice.

Miepee commented on randovania/games/prime1/pickup_database/pickup-database.json:
```diff
@@ -311,7 +317,8 @@
             "model_name": "Morph Ball",
```
you have a boost ball sprite for echoes assigned, but not for prime?

UltiNaruto replied:
I kinda forgot and since offworld models are removed for now, it's no more an issue

UltiNaruto commented on randovania/games/planets_zebeth/assets/hash_words.json:
```diff
@@ -0,0 +1,65 @@
+    "Spark",
+    "Squeept",
+    "Statue",
+    "Tangle",
```
Since you're selecting 4 lines could you precise which line it is?

Miepee replied:
Github just shows the context of the previous line
Line in question is Tangle.

Miepee commented on test/games/planets_zebeth/layout/test_planets_zebeth_describer.py:
```diff
@@ -0,0 +1,59 @@
+    preset = PresetManager(None).default_preset_for_game(RandovaniaGame.METROID_PLANETS_ZEBETH).get_preset()
+    assert isinstance(preset.configuration, PlanetsZebethConfiguration)
+
+    if not artifacts.vanilla_tourian_keys and artifacts.required_artifacts != 2:
+        # Remove or add missile expansions if Tourian key count is not 2
+        for ammo_def, ammo in preset.configuration.ammo_pickup_configuration.pickups_state.items():
+            if ammo_def.name == "Missile Tank":
+                ammo_ = dataclasses.replace(
+                    ammo,
+                    pickup_count=ammo.pickup_count - (artifacts.required_artifacts - 2),
+                )
+                preset.configuration.ammo_pickup_configuration.pickups_state[ammo_def] = ammo_
+                break
```
as said in discord, i'm really not a fan of this condition. either only test artifacts <= 2, or always modify the preset to have ~10 missiles less.

UltiNaruto replied:
can't I use a second preset with 9 keys?

Miepee replied:
If there's a good reason for you to ship a second preset to the end user where the only modification is "9 keys and adjusted items" sure. Otherwise, you can also just use the preset as a test file and load that, like some other tests do.

UltiNaruto replied:
tbf I wanted to provide another starter preset that uses non vanilla keys

UltiNaruto replied:
changed it to use a second preset

Miepee reviewed (commented):
Only maintainers needed now.

Miepee commented:
unsure if you noticed, but tests are broken.

UltiNaruto commented:
Looks like I'm missing a migration_data.json file in assets folder

UltiNaruto commented:
> unsure if you noticed, but tests are broken.

Tests are fixed

Miepee commented on randovania/games/planets_zebeth/pickup_database/pickup-database.json:
```diff
@@ -0,0 +1,298 @@
+            "pickup_category": "beam",
+            "broad_category": "beam_related",
+            "model_name": "spr_ITEM_Long_Beam",
+            "offworld_models": {},
```
am2r has long beam now

UltiNaruto replied:
what's the model name for long beam?

Miepee replied:
sItemLongBeam

Miepee commented on randovania/games/planets_zebeth/pickup_database/pickup-database.json:
```diff
@@ -0,0 +1,298 @@
+            "model_name": "spr_ITEM_Ice_Beam",
+            "offworld_models": {
+                "prime1": "Ice Beam",
+                "dread": "powerup_icemissile",
```
looking at other pickup DBs, pretty sure dunc would prefer to not have this for dread

Miepee commented on randovania/games/planets_zebeth/pickup_database/pickup-database.json:
```diff
@@ -0,0 +1,298 @@
+            "model_name": "spr_ITEM_Missile",
+            "offworld_models": {
+                "prime1": "Missile",
+                "prime2": "MissileLauncher",
```
iirc for echoes, you should use MissileExpansionPrime1

Miepee commented on randovania/games/planets_zebeth/pickup_database/pickup-database.json:
```diff
@@ -0,0 +1,298 @@
+            "preferred_location_category": "major",
+            "original_location": 16
+        },
+        "Ice Beam": {
```
you can now add the original location for ice beam

UltiNaruto replied:
gotta wait till the branch for multiple original locations is merged

Miepee replied:
oh whoops

Miepee commented on randovania/games/planets_zebeth/pickup_database/pickup-database.json:
```diff
@@ -0,0 +1,298 @@
+            "hinted_as_major": false
+        }
+    },
+    "standard_pickups": {
```
afaik you need to update the version of your pickup db; original_location is a list nowadays.

UltiNaruto replied:
it's not a thing yet

UltiNaruto commented:
Anything else to fix or change?

ThanatosGit commented on randovania/games/planets_zebeth/gui/dialog/cosmetic_patches_dialog.py:
```diff
@@ -0,0 +1,41 @@
+
+        assert isinstance(current, PlanetsZebethCosmeticPatches)
+        self._cosmetic_patches = current
+
```
As the selector is already in the GUI, I suggest to also fill it and to not have an empty selector.
```suggestion
        for room_gui_type in PlanetsZebethRoomGuiType:
            self.room_name_dropdown.addItem(room_gui_type.long_name, room_gui_type)
```

ThanatosGit commented on randovania/games/planets_zebeth/game_data.py:
```diff
@@ -0,0 +1,93 @@
+        (
+            "Which versions of Metroid Planets are supported?",
+            "Only version 1.27g is supported. "
+            "Later versions are embedding code in the executable "
+            "which prevents modifying the code.",
```
I don't think we need such technical details for the users.
As this gets added with the development state first, I also wouldn't go to crazy with polishing here because I assume there will be more entries later anyways. So feel free to ignore this comment for now.

UltiNaruto replied:
I might simplify it by saying "Any other versions will fail to export"?

ThanatosGit replied:
Sounds good

UltiNaruto replied:
Done

ThanatosGit reviewed (commented):
Only have those two short comments, otherwise I think it's really enough for the development state.

Miepee reviewed (approved):
Don't see anything else that stands out. Let's try to get this in.
